numpyro/infer/reparam.py

```diff
@@ -30,23 +29,23 @@ def __call__(self, name, fn, obs):
     def _unwrap(self, fn):
         """
         Unwrap Independent(...) and ExpandedDistribution(...) distributions.
+        We can recover the input `fn` from the result triple `(fn, expand_shape, event_dim)`
+        with `fn.expand(expand_shape).to_event(event_dim - fn.event_dim)`.
         """
-        batch_shape = fn.batch_shape
+        shape = fn.shape()
         event_dim = fn.event_dim
         while isinstance(fn, (dist.Independent, dist.ExpandedDistribution)):
             fn = fn.base_dist
-        return fn, batch_shape, event_dim
+        expand_shape = shape[:len(shape) - fn.event_dim]
+        return fn, expand_shape, event_dim
 
-    def _wrap(self, fn, batch_shape, event_dim):
+    def _wrap(self, fn, expand_shape, event_dim):
         """
         Wrap in Independent and ExpandedDistribution distributions.
         """
         # Match batch_shape.
         assert fn.event_dim <= event_dim
-        fn_batch_shape = batch_shape + (1,) * (event_dim - fn.event_dim)
-        fn_batch_shape = lax.broadcast_shapes(fn_batch_shape, fn.batch_shape)
-        if fn.batch_shape != fn_batch_shape:
-            fn = fn.expand(fn_batch_shape)
+        fn = fn.expand(expand_shape)  # no-op if expand_shape == fn.batch_shape
```

Comment:
@fritzo This is a bit subtle. Currently, when we unwrap `dist.Normal(0, 1).expand([10]).to_event()`, the `fn, batch_shape, event_dim` is `dist.Normal(0, 1), (), 1`. So we are losing the shape `(10,)` information. I have adjusted the logic a bit such that `_unwrap` will return `fn, expand_shape, event_dim` such that we can always recover the original distribution with `fn.expand(expand_shape).to_event(event_dim - fn.event_dim)`. With that, it seems to me that we don't need this broadcasting logic. As a trade-off, we have to make sure that regardless how we reparam the base `fn`, event dimensions need to be consistent, that is
```
fn, expand_shape, event_dim = _unwrap(fn)
reparam_fn = foo(fn)
assert foo_inv(reparam_fn).event_dim == fn.event_dim
reparam_event_dim = event_dim - fn.event_dim + reparam_fn.event_dim
base_value = sample(name, _wrap(reparam_fn, expand_shape, reparam_event_dim))
```
For example, in ProjectedNormal, reparam_fn should be `dist.Normal(...).to_event(1)`. Similarly, for TransformReparam, we need to make sure that `transform(base_dist).event_dim == transformed_dist.event_dim`, i.e. an atom event of base_dist should be transformed exactly to an atom event of transformed_dist. This is guaranteed given our recent change in TransformedDistribution.

Reply:
Indeed that's subtle, thanks for explaining.